### PR TITLE
EnGe1 OK and documented, document `z_snap`

### PR DESCRIPTION
**Pull request by EllipticEllipsis**
Sadly this actor was gutted into a cutscene actor in MM, but for some reason it is still registerable as pictographable. This sent me down a rabbit-hole to document `z_snap` and related actors. Most of the object is identical to OoT with the exception of the new animations.

I haven't made a header for snap yet, can do so if we'd rather do that now. I'm not enormously happy with some of the names, particularly that last function and the 0x144 function pointer functions that use it, open to suggestions.

Of course upgrading the other actors to PictoActor (which I think is a reasonable name for the struct) required a lot of reformatting too...

### CODE REVIEW

File: assets/xml/objects/object_ge1.xml
Comment (AngheloAlf) [inline assets/xml/objects/object_ge1.xml:35]: Any reason to not name this `gGerudoWhiteSkel`?

File: src/overlays/actors/ovl_En_Bal/z_en_bal.h
Comment (AngheloAlf) [inline src/overlays/actors/ovl_En_Bal/z_en_bal.h:12]:
```suggestion
    /* 0x000 */ PictoActor picto;
    /* 0x148 */ char unk_148[0x44];
```

File: src/overlays/actors/ovl_En_Mnk/z_en_mnk.h
Comment (AngheloAlf) [inline src/overlays/actors/ovl_En_Mnk/z_en_mnk.h:12]:
```suggestion
    /* 0x000 */ PictoActor picto;
    /* 0x148 */ char unk_148[0x2A0];
```

File: assets/xml/objects/object_ge1.xml
Comment (EllipticEllipsis) [reply assets/xml/objects/object_ge1.xml:35]: Oh, missed the obvious, whoops.

File: src/overlays/actors/ovl_Obj_Grass/z_obj_grass.c
Comment (AngheloAlf) [inline src/overlays/actors/ovl_Obj_Grass/z_obj_grass.c:10]: I think including this header _after_ the others doesn't do anything, could be wrong tho. Could you check here?

File: src/overlays/actors/ovl_Obj_Hunsui/z_obj_hunsui.c
Comment (AngheloAlf) [inline src/overlays/actors/ovl_Obj_Hunsui/z_obj_hunsui.c:7]: remove comment

File: src/overlays/actors/ovl_En_Bigokuta/z_en_bigokuta.c
Comment (engineer124) [inline src/overlays/actors/ovl_En_Bigokuta/z_en_bigokuta.c:183]:
```suggestion
    return Snap_ValidatePictograph(play, thisx, PICTOGRAPH_BIG_OCTO, &thisx->focus.pos, &thisx->shape.rot, 280.0f, 1800.0f, -1);
```

File: src/overlays/actors/ovl_En_Zov/z_en_zov.c
Comment (engineer124) [inline src/overlays/actors/ovl_En_Zov/z_en_zov.c:468]:
```suggestion
    ret = Snap_ValidatePictograph(play, &this->picto.actor, PICTOGRAPH_LULU_HEAD, &this->picto.actor.focus.pos,
                                  &this->picto.actor.shape.rot, 10.0f, 300.0f, -1);
    ret |= Snap_ValidatePictograph(play, &this->picto.actor, PICTOGRAPH_LULU_RIGHT_ARM, &this->unk_308, &this->picto.actor.shape.rot, 50.0f,
                                   160.0f, 0x3000);
    ret |= Snap_ValidatePictograph(play, &this->picto.actor, PICTOGRAPH_LULU_LEFT_ARM, &this->unk_314, &this->picto.actor.shape.rot, 50.0f,
                                   160.0f, 0x3000);
```

File: src/overlays/actors/ovl_Obj_Grass/z_obj_grass.c
Comment (EllipticEllipsis) [reply src/overlays/actors/ovl_Obj_Grass/z_obj_grass.c:10]: Nope, definitely required. And I don't see why it wouldn't do anything.

**Review by AngheloAlf — APPROVED** [review]

File: src/overlays/actors/ovl_En_Ge1/z_en_ge1.c
Comment (hensldm) [inline src/overlays/actors/ovl_En_Ge1/z_en_ge1.c:86]:
```suggestion
    SkelAnime_InitFlex(play, &this->skelAnime, &gGerudoWhiteSkel, &gGerudoWhiteArmsFoldedAnim, this->jointTable,
                       this->morphTable, GERUDO_WHITE_LIMB_MAX);
```

File: src/overlays/actors/ovl_En_Ge1/z_en_ge1.c
Comment (hensldm) [inline src/overlays/actors/ovl_En_Ge1/z_en_ge1.c:53]:
```suggestion
#define GERUDO_WHITE_STATE_DISABLE_MOVEMENT (1 << 3)
```

File: src/overlays/actors/ovl_En_Ge1/z_en_ge1.c
Comment (hensldm) [inline src/overlays/actors/ovl_En_Ge1/z_en_ge1.c:110]: Could probably make an enum for the type as well.

File: assets/xml/objects/object_ge1.xml
Comment (engineer124) [inline assets/xml/objects/object_ge1.xml:12]: Why maybe?

File: src/code/z_snap.c
Comment (engineer124) [inline src/code/z_snap.c:88]: Except it's also uesd above in `Snap_RecordPictographedActors`

File: include/z64save.h
Comment (engineer124) [inline include/z64save.h:146]:
```suggestion
    /* 0x0E78 */ u32 pictoSuccessFlags;                       // Flags set by `PictoActor`s if pictograph is valid
    /* 0x0E7C */ u32 pictoErrorFlags;                       // Flags set by Snap_ValidatePictograph() to record errors; volatile since that function is run many times in succession
```
I wonder if we could give these more interesting names now, and leave it to comments that the error flag uses `0x20` to indicate that it's an error (should that get a define?)

File: src/code/z_snap.c
Comment (engineer124) [inline src/code/z_snap.c:189]:
```suggestion
    // If all of the above checks pass, set the flag
    if (errorFlags == 0) {
```
Perhaps?

File: src/code/z_snap.c
Comment (engineer124) [inline src/code/z_snap.c:77]:
```suggestion
                    if ((pictoActor->validationFunc)(play, actor) == 0) {
                        validCount++;
```
Or something similar?

File: src/code/z_snap.c
Comment (engineer124) [inline src/code/z_snap.c:168]:
```suggestion
    x = (s16)(projectedPos.x * distance * (SCREEN_WIDTH / 2) + (SCREEN_WIDTH / 2)) - ((SCREEN_WIDTH / 2) + 5);
    y = (s16)(projectedPos.y * distance * -(SCREEN_HEIGHT / 2) + (SCREEN_HEIGHT / 2)) - ((SCREEN_HEIGHT / 2) + 7);
    if ((x < 0) || (x > (SCREEN_WIDTH - 10)) || (y < 0) || (y > (SCREEN_HEIGHT - 15))) {
```
Might as well go the full way t onot worry about conflicts. Some of these literals seem weird though, specifically the shift of `85` and `67`. Do you think my interpretation of those constants are right?

Note: without the shifts, you directly get screen coordinates
```
x = (s16)(projectedPos.x * distance * (SCREEN_WIDTH / 2) + (SCREEN_WIDTH / 2));
y = (s16)(projectedPos.y * distance * -(SCREEN_HEIGHT / 2) + (SCREEN_HEIGHT / 2));
```

File: src/code/z_snap.c
Comment (engineer124) [inline src/code/z_snap.c:13]: When looping through an entire enum from 0 - max, I tend to leave it as 0 and not the enum, because "0" is more important than the value of the enum itself (i.e. shifting `ACTORCAT_SWITCH` to a different value shouldn't shift this value away from 0)

Comment (EllipticEllipsis) [reply src/code/z_snap.c:13]: I'd be inclined to add `ACTORCAT_MIN` for such situations.

File: include/z64save.h
Comment (EllipticEllipsis) [reply include/z64save.h:146]: I think the thing is that there isn't really any reason to use one over the other for each thing, it just happens to use the topmost few of the second one for error output.

File: assets/xml/objects/object_ge1.xml
Comment (EllipticEllipsis) [reply assets/xml/objects/object_ge1.xml:12]: @tom-overton wasn't completely sure about this correspondence.

File: src/overlays/actors/ovl_En_Ge1/z_en_ge1.c
Comment (EllipticEllipsis) [reply src/overlays/actors/ovl_En_Ge1/z_en_ge1.c:110]: I don't think there's much point here, it's used once, only one is used, and it doesn't really do anything (the OoT ones have lots of different ones that set the same hairstyle: https://github.com/zeldaret/oot/blob/9647f50451f26fa8b96653bf792f86c6148f6d12/src/overlays/actors/ovl_En_Ge1/z_en_ge1.c#L107-L173 )

File: assets/xml/objects/object_ge1.xml
Comment (EllipticEllipsis) [reply assets/xml/objects/object_ge1.xml:35]: Usually this is singular: the enum type refers to one of the thing.

File: src/code/z_snap.c
Comment (engineer124) [reply src/code/z_snap.c:13]: Iterating over an enum from 0 to max is ubiquitous over the repo, would you propose a `_MIN` for all those enums?

File: assets/xml/objects/object_ge1.xml
Comment (EllipticEllipsis) [reply assets/xml/objects/object_ge1.xml:12]: These ones I think I believe though, it's the first one that's a bit strange.

File: src/code/z_snap.c
Comment (EllipticEllipsis) [reply src/code/z_snap.c:77]: Yeah, although I'm not sure I really like `valid`, it was the best I could think of. At least if they're consistent they're easy to change.

File: src/code/z_snap.c
Comment (EllipticEllipsis) [reply src/code/z_snap.c:168]: As discussed in Discord, I'll make this a TODO rather than attempting to nail it down now.

File: assets/xml/objects/object_ge1.xml
Comment (hensldm) [reply assets/xml/objects/object_ge1.xml:35]: I don't remember seeing singular ones around, I think plural is more used. Though now that you mention it, singular probably is better.

File: src/code/z_snap.c
Comment (EllipticEllipsis) [reply src/code/z_snap.c:13]: Yes, I've been thinking about this for a while. You'd do the same for different sections of enums that are made out of chunks, like masks in the item enum (or a less bad example...)

File: src/code/z_snap.c
Comment (EllipticEllipsis) [reply src/code/z_snap.c:189]: I think ret's fine, this is pretty much how you usually use ret. It's also not returning flags anyway.

File: src/overlays/actors/ovl_En_Ge1/z_en_ge1.c
Comment (hensldm) [reply src/overlays/actors/ovl_En_Ge1/z_en_ge1.c:110]: I think it is still valuable. It would be the outward facing enum that someone could use to spawn each type.

File: src/overlays/actors/ovl_En_Ge1/z_en_ge1.c
Comment (hensldm) [inline src/overlays/actors/ovl_En_Ge1/z_en_ge1.c:150]: 8 and 9 here could use the anim enum

File: src/overlays/actors/ovl_En_Ge1/z_en_ge1.c
Comment (EllipticEllipsis) [reply src/overlays/actors/ovl_En_Ge1/z_en_ge1.c:110]: And call them what? USED, UNUSED_1, UNUSED_2?

File: src/code/z_snap.c
Comment (EllipticEllipsis) [reply src/code/z_snap.c:13]: Would look like
```c
typedef enum {
    /* 0x00 */ ACTORCAT_MIN,
    /* 0x00 */ ACTORCAT_SWITCH = ACTORCAT_MIN,
    /* 0x01 */ ACTORCAT_BG,
    /* 0x02 */ ACTORCAT_PLAYER,
    /* 0x03 */ ACTORCAT_EXPLOSIVES,
    /* 0x04 */ ACTORCAT_NPC,
    /* 0x05 */ ACTORCAT_ENEMY,
    /* 0x06 */ ACTORCAT_PROP,
    /* 0x07 */ ACTORCAT_ITEMACTION,
    /* 0x08 */ ACTORCAT_MISC,
    /* 0x09 */ ACTORCAT_BOSS,
    /* 0x0A */ ACTORCAT_DOOR,
    /* 0x0B */ ACTORCAT_CHEST,
    /* 0x0C */ ACTORCAT_MAX
} ActorType;
```
but I won't do it for now.

File: src/overlays/actors/ovl_En_Ge1/z_en_ge1.c
Comment (hensldm) [reply src/overlays/actors/ovl_En_Ge1/z_en_ge1.c:110]: Personally I would just name them after their hairstyle, since that seems to distinguish them all.

File: src/code/z_snap.c
Comment (engineer124) [reply src/code/z_snap.c:189]: I've never been a fan of `ret` to begin with, it's only a slight step up from the generic `temp` imo. It's appropriate in libultra where the goal is to retain names that the original devs used, fair enough. But if there's something meaningful to the ret, it should be named (might just be me though).

Is it not returning flags? It's returning values from the `PictographFlag` enum, specifically the error flags with the `0x20` involved

Comment (EllipticEllipsis) [reply src/code/z_snap.c:189]: It returns an or'd-together mess that you can't interpret apart from "it failed"

Comment (engineer124) [reply src/code/z_snap.c:189]: Actually yeah you're right. It's confusing because the enum isn't the flags themselves, it's the index for the shift of the flags i.e. `1 << enumVal`, but it's `|= enumVal` in this return. Fair enough. I wonder if `PictographFlag` is fine since it's not really a flag, but the shift value for a flag. It's probably fine though

**Review by hensldm — APPROVED** [review]

File: src/code/z_snap.c
Comment (engineer124) [inline src/code/z_snap.c:170]:
```suggestion
    #define PICTO_SCREEN_WIDTH 150
    #define PICTO_SCREEN_HEIGHT 105
    
    Actor_GetProjectedPos(play, pos, &projectedPos, &distance);
    
   // Get the screen coordinates, then shift the origin to the top left corner of the smaller interior pictoscreen
    x = (s16)(projectedPos.x * distance * (SCREEN_WIDTH / 2) + (SCREEN_WIDTH / 2)) - ((SCREEN_WIDTH 
 - PICTO_SCREEN_WIDTH) / 2);
    y = (s16)(projectedPos.y * distance * -(SCREEN_HEIGHT / 2) +(SCREEN_HEIGHT / 2)) - ((SCREEN_HEIGHT 
 - PICTO_SCREEN_HEIGHT) / 2);
 
    // checks if the coordinates are within the sub-screen of the picto-photo
    if ((x < 0) || (x > PICTO_SCREEN_WIDTH) || (y < 0) || (y > PICTO_SCREEN_HEIGHT)) {
```
Yeah, your assumption was right. This gets the screen coordinates, shifts the origin of the coordinates to the top-left corner of that smaller rectangle screen for the pictograph, than checks if it's within that smaller screen.

An alternative define would be to `#define PICTO_SCREEN_CUTOFF 85` define the dimensions of the screen that's cutoff, instead of the dimensions of the interior picto size, but that seems less intuitive

Comment (engineer124) [reply src/code/z_snap.c:170]: So I wonder if `PICTOGRAPH_NOT_IN_VIEW` should be `PICTOGRAPH_NOT_IN_PICTO_SCREEN` or `PICTOGRAPH_NOT_IN_SUB_SCREEN` or something?

Comment (EllipticEllipsis) [reply src/code/z_snap.c:170]: `NOT_IN_PICTO_CAPTURE_AREA`?

Comment (EllipticEllipsis) [reply src/code/z_snap.c:170]: And it's more that it's the top-left corner of the capture region for the initial shifting. I'll think about it myself, see what I come up with.

File: src/code/z_snap.c
Comment (EllipticEllipsis) [reply src/code/z_snap.c:191]: I think it's consistent with what we do elsewhere for things like the flag functions in z_actor, where the `flag` input is the shift/index. But this is probably precisely what confused them in the first place, and it does make it very inconvenient to find out how a pictograph failed, since the actual flags in SaveContext are volatile and change value every time an actor runs this function. I'll add to the note, and we might want to consider how to differentiate these in future.

File: src/code/z_snap.c
Comment (EllipticEllipsis) [reply src/code/z_snap.c:170]: Would `RESOLUTION` be a reasonable description, or is that unrelated to the actual screen coordinates?

Comment (Kenix3) [reply src/code/z_snap.c:170]: Resolution could work for the actual picto frame. Would also prefer not name it screen width or height. Capture area or some other suggestion like that is better. I don't think it needs to hold up merging because you stated documentation is not totally complete.

**Review by Kenix3 — APPROVED** [review]

**Comment by EllipticEllipsis** [comment]
Following discussion in discord, updated quite a lot of stuff, recommend re-review.

**Review by hensldm — APPROVED** [review]
I had gone through while it was building. Changes look good to me.

File: src/code/z_actor.c
Comment (engineer124) [inline src/code/z_actor.c:399]: What's the motivation for `invW` instead of `cappedInvW`? Note that many of these changes here directly clash with https://github.com/zeldaret/mm/pull/933 so we should try to sync these

File: include/macros.h
Comment (engineer124) [inline include/macros.h:15]: From my understanding, these macros goes against the current naming scheme of the oot/mm repos. I'm not necessarily opposed to the changed, but if I understand correctly, this makes the assumption that:
`projectionPos` is renamed to `screenPos`
`screenPos`  is renamed to `devicePos`


`projectionPos`/`screenPos` has the advantage of being more intuitive and more standard (projectionPos based on projectionMtx), and seems easier to understand from users (both projection and screen positions make sense to me, screen/device did not make intuitive sense at first).
Plus, everyone is used to screen coordinates being (0, SCREEN_WIDTH), instead of what you propose here (-1, 1)

On the other hand, there does seem to be consistency with Nintendo for the change you're proposing:
https://ultra64.ca/files/documentation/online-manuals/man-v5-2/allman52/kantan/step3/1/1_3.htm

Many things would need to change to follow this convention you're proposing (all the projected pos, `SCREEN_XXX` macros, it could quickly add up to 1000's of lines)

I don't feel the need to follow Ninentdo's naming convension at all if there's better and less confusing names. but I would be open to hearing others feedback, and I think this is a bigger discussion to be had instead of tacking on to the end of a PR

File: src/code/z_actor.c
Comment (EllipticEllipsis) [reply src/code/z_actor.c:399]: I don't think it being capped is a crucial part of the operation, rather than just avoiding dividing by something small.

File: include/macros.h
Comment (EllipticEllipsis) [reply include/macros.h:15]: I'll put it back to `PROJECTED_TO_SCREEN` for now, then.

**Pull request merged**
